packages/airnode-protocol/contracts/rrp/requesters/RrpBeaconServer.sol

```diff
@@ -242,24 +246,39 @@ contract RrpBeaconServer is
         bytes32 templateId = requestIdToTemplateId[requestId];
         require(templateId != bytes32(0), "No such request made");
         delete requestIdToTemplateId[requestId];
-        int256 decodedData = abi.decode(data, (int256));
+        (int256 decodedData, uint256 decodedTimestamp) = abi.decode(
+            data,
+            (int256, uint256)
+        );
         require(
             decodedData >= type(int224).min && decodedData <= type(int224).max,
             "Value typecasting error"
         );
         require(
-            block.timestamp <= type(uint32).max,
+            decodedTimestamp <= type(uint32).max,
             "Timestamp typecasting error"
         );
+        require(
+            decodedTimestamp > templateIdToBeacon[templateId].timestamp,
+            "Fulfillment older than beacon"
+        );
+        require(
+            decodedTimestamp + 1 hours > block.timestamp,
+            "Fulfillment stale"
+        );
+        require(
+            decodedTimestamp - 1 hours < block.timestamp,
+            "Fulfillment from future"
+        );
```

Comment:
The tolerance values are not strict (+- 1 hour) assuming that this is not a significant time period in the context of asset prices, and stricter checks are more vulnerable (https://consensys.github.io/smart-contract-best-practices/recommendations/#timestamp-dependence) with no positive tradeoff. The general rule for Ethereum is that you can trust the chain timestamp up to 15 seconds, but note that this contract will also be used in other chains whose node implementation may not be as strict/where consensus attacks are easier to pull off.

Reply:
I'm not sure how to feel about having the +- 1 being hardcoded but the reasoning is there so I guess that suffices. Is there any practical case in which we get  decodedTimestamp - 1 hours < block.timestamp ?

Reply:
That shouldn't happen normally, but the consequences are permanent if it happens (https://github.com/api3dao/airnode/pull/695#discussion_r757021322) so it needs to be accounted for